ring/util.go

```diff
@@ -0,0 +1,222 @@
+			currRingState, _ := r.GetAllHealthy(op) // nolint:errcheck
+
+			if HasReplicationSetChanged(ringLastState, currRingState) {
+				ringLastState = currRingState
+				ringLastStateTs = time.Now()
+			} else if time.Since(ringLastStateTs) >= minStability {
+				return nil
+			}
+		}
+	}
+}
+
+// MakeBuffersForGet returns buffers to use with Ring.Get().
+func MakeBuffersForGet() (bufDescs []InstanceDesc, bufHosts, bufZones []string) {
+	bufDescs = make([]InstanceDesc, 0, GetBufferSize)
+	bufHosts = make([]string, 0, GetBufferSize)
+	bufZones = make([]string, 0, GetBufferSize)
+	return
+}
+
+// getZones return the list zones from the provided tokens. The returned list
+// is guaranteed to be sorted.
+func getZones(tokens map[string][]uint32) []string {
+	var zones []string
+
+	for zone := range tokens {
+		zones = append(zones, zone)
+	}
+
+	sort.Strings(zones)
+	return zones
+}
+
+// searchToken returns the offset of the tokens entry holding the range for the provided key.
+func searchToken(tokens []uint32, key uint32) int {
+	i := sort.Search(len(tokens), func(x int) bool {
+		return tokens[x] > key
+	})
+	if i >= len(tokens) {
+		i = 0
+	}
+	return i
+}
+
+// GetFirstAddressOf returns the first IPv4 address of the supplied interface names, omitting any 169.254.x.x automatic private IPs if possible.
+func getFirstAddressOf(names []string, logger log.Logger) (string, error) {
+	var ipAddr string
+	for _, name := range names {
+		inf, err := net.InterfaceByName(name)
+		if err != nil {
+			level.Warn(logger).Log("msg", "error getting interface", "inf", name, "err", err)
+			continue
+		}
+		addrs, err := inf.Addrs()
+		if err != nil {
+			level.Warn(logger).Log("msg", "error getting addresses for interface", "inf", name, "err", err)
+			continue
+		}
+		if len(addrs) <= 0 {
+			level.Warn(logger).Log("msg", "no addresses found for interface", "inf", name, "err", err)
+			continue
+		}
+		if ip := filterIPs(addrs); ip != "" {
+			ipAddr = ip
+		}
+		if strings.HasPrefix(ipAddr, `169.254.`) || ipAddr == "" {
+			continue
+		}
+		return ipAddr, nil
+	}
+	if ipAddr == "" {
+		return "", fmt.Errorf("No address found for %s", names)
+	}
+	if strings.HasPrefix(ipAddr, `169.254.`) {
+		level.Warn(logger).Log("msg", "using automatic private ip", "address", ipAddr)
+	}
+	return ipAddr, nil
+}
+
+// filterIPs attempts to return the first non automatic private IP (APIPA / 169.254.x.x) if possible, only returning APIPA if available and no other valid IP is found.
+func filterIPs(addrs []net.Addr) string {
+	var ipAddr string
+	for _, addr := range addrs {
+		if v, ok := addr.(*net.IPNet); ok {
+			if ip := v.IP.To4(); ip != nil {
+				ipAddr = v.IP.String()
+				if !strings.HasPrefix(ipAddr, `169.254.`) {
+					return ipAddr
+				}
+			}
+		}
+	}
+	return ipAddr
+}
+
+// StringsContain returns true if the search value is within the list of input values.
+func StringsContain(values []string, search string) bool {
```

Comment:
Since we use StringsContain in client/pool.go removeStaleClients we have to expose it here. I can see if it helps to move it to a folder within the package.